File: HTTPShortcuts/curl_command/src/main/java/ch/rmy/curlcommand/CurlParser.java
package ch.rmy.curlcommand;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;
import java.util.ListIterator;

public class CurlParser {

    public static CurlCommand parse(String commandString) {
        List<String> arguments = CommandParser.parseCommand(commandString);
        CurlParser curlParser = new CurlParser(arguments);
        return curlParser.builder.build();
    }

    private final CurlCommand.Builder builder = new CurlCommand.Builder();

    private CurlParser(List<String> arguments) {
        ListIterator<String> iterator = arguments.listIterator();
        boolean urlFound = false;
        while (iterator.hasNext()) {
            int index = iterator.nextIndex();
            String argument = iterator.next();

            if (argument.equalsIgnoreCase("curl") && index == 0) {
                continue;
            }

            if (argument.startsWith("-") && !argument.startsWith("--") && argument.length() > 2) {
                String argumentParameter = argument.substring(2);
                argument = argument.substring(0, 2);
                iterator.add(argumentParameter);
                iterator.previous();
            }

            // arguments with 1 parameter
            if (iterator.hasNext()) {
                switch (argument) {
                    case "-X":
                    case "--request": {
                        builder.method(iterator.next());
                        continue;
                    }
                    case "-H":
                    case "--header": {
                        String[] header = iterator.next().split(": ", 2);
                        if (header.length == 2) {
                            builder.header(header[0], header[1]);
                        }
                        continue;
                    }
                    case "-d":
                    case "--data":
                    case "--data-binary":
                    case "--data-urlencode": {
                        String data = iterator.next();
                        if (argument.equals("--data-urlencode")) {
                            try {
                                if (data.contains("=")) {
                                    String[] parts = data.split("=", 2);
                                    data = parts[0] + "=" + URLEncoder.encode(parts[1], "utf-8");
                                } else {
                                    data = URLEncoder.encode(data, "utf-8");
                                }
                            } catch (UnsupportedEncodingException e) {
                            }
                        }
                        builder.data(data);
                        continue;
                    }
                    case "-m":
                    case "--max-time":
                    case "--connect-timeout": {
                        try {
                            builder.timeout(Integer.parseInt(iterator.next()));
                        } catch (NumberFormatException e) {
                        }
                        continue;
                    }
                    case "-u":
                    case "--user": {
                        String[] credentials = iterator.next().split(":", 2);
                        builder.username(credentials[0]);
                        if (credentials.length > 1) {
                            builder.password(credentials[1]);
                        }
                        continue;
                    }
                    case "-A":
                    case "--user-agent": {
                        builder.header("User-Agent", iterator.next());
                        continue;
                    }
                    case "--url": {
                        String url = iterator.next();
                        if (url.toLowerCase().startsWith("http")) {
                            builder.url(url);
                        } else {
                            builder.url("http://" + url);
                        }
                        continue;
                    }
                    case "-e":
                    case "--referer": {
                        builder.header("Referer", iterator.next());
                        continue;
                    }
                }
            }

            // arguments with 0 parameters
            switch (argument) {
                case "-G":
                case "--get": {
                    builder.method(CurlCommand.METHOD_GET);
                    continue;
                }
            }

            if (argument.toLowerCase().startsWith("http")) {
                urlFound = true;
                builder.url(argument);
            } else if (!argument.startsWith("-") && !urlFound) {
                urlFound = true;
                builder.url("http://" + argument);
            }
        }
    }

}

File: HTTPShortcuts/curl_command/src/main/kotlin/ch/rmy/curlcommand/CurlParser.kt
package ch.rmy.curlcommand

import java.net.URLEncoder

class CurlParser private constructor(arguments: List<String>) {

    private val builder = CurlCommand.Builder()

    init {
        val args = arguments.toMutableList()
        val iterator = args.listIterator()
        var urlFound = false
        var wasNoParamArgument = false
        var spaceInjected = false
        loop@ while (iterator.hasNext()) {
            val index = iterator.nextIndex()
            var argument = iterator.next()

            if (argument.equals("curl", ignoreCase = true) && index == 0) {
                continue
            }

            if (wasNoParamArgument && spaceInjected) {
                argument = "-$argument"
            }

            if (argument.startsWith("-") && !argument.startsWith("--") && argument.length > 2) {
                val argumentParameter = argument.substring(2)
                argument = argument.substring(0, 2)
                spaceInjected = true
                iterator.add(argumentParameter)
                iterator.previous()
            } else {
                spaceInjected = false
            }

            wasNoParamArgument = false

            // arguments with 1 parameter
            if (iterator.hasNext()) {
                when (argument) {
                    "-X", "--request" -> {
                        builder.method(iterator.next())
                        continue@loop
                    }
                    "-x", "--proxy" -> {
                        val parts = iterator.next().split(":")
                        builder.proxy(parts[0], parts.getOrNull(1)?.toIntOrNull() ?: 3128)
                        continue@loop
                    }
                    "-H", "--header" -> {
                        val header = iterator.next().split(":", limit = 2)
                        if (header.size == 2) {
                            builder.header(header[0], header[1].removePrefix(" "))
                        }
                        continue@loop
                    }
                    "-d", "--data", "--data-binary", "--data-urlencode", "--data-raw" -> {
                        builder.methodIfNotYetSet("POST")
                        var dataItems = iterator.next().split("&")
                        if (argument == "--data-urlencode") {
                            dataItems = dataItems.map { data ->
                                if (data.contains("=")) {
                                    val parts = data.split("=", limit = 2)
                                    parts[0] + "=" + URLEncoder.encode(parts[1], "utf-8")
                                } else {
                                    URLEncoder.encode(data, "utf-8")
                                }
                            }
                        }
                        if (argument == "--data-binary" && dataItems.any { it.startsWith("@") }) {
                            builder.usesBinaryData()
                        } else {
                            dataItems.forEach { data ->
                                builder.data(data)
                            }
                        }
                        continue@loop
                    }
                    "-F", "--form" -> {
                        val data = iterator.next()
                        builder.isFormData()
                        builder.data(data)
                        builder.methodIfNotYetSet("POST")
                        continue@loop
                    }
                    "-m", "--max-time", "--connect-timeout" -> {
                        iterator.next()
                            .toIntOrNull()
                            ?.let {
                                builder.timeout(it)
                            }
                        continue@loop
                    }
                    "-u", "--user" -> {
                        val credentials = iterator.next().split(":", limit = 2)
                        builder.username(credentials[0])
                        if (credentials.size > 1) {
                            builder.password(credentials[1])
                        }
                        continue@loop
                    }
                    "--digest" -> {
                        builder.isDigestAuth()
                        continue@loop
                    }
                    "-A", "--user-agent" -> {
                        builder.header("User-Agent", iterator.next())
                        continue@loop
                    }
                    "--url" -> {
                        builder.url(iterator.next())
                        continue@loop
                    }
                    "-e", "--referer" -> {
                        builder.header("Referer", iterator.next())
                        continue@loop
                    }
                }
            }

            // arguments with 0 parameters
            wasNoParamArgument = argument.startsWith("-") && !argument.startsWith("--")
            when (argument) {
                "-G", "--get" -> {
                    builder.forceGet()
                    continue@loop
                }
                "-k", "--insecure" -> {
                    builder.insecure()
                    continue@loop
                }
                "-s", "--silent" -> {
                    builder.silent()
                    continue@loop
                }
                "-I", "--head" -> {
                    builder.method("HEAD")
                    continue@loop
                }
                "-4", "--ipv4" -> {
                    builder.ipVersion4()
                    continue@loop
                }
                "-6", "--ipv6" -> {
                    builder.ipVersion6()
                    continue@loop
                }
            }
            wasNoParamArgument = false

            if (argument.startsWith("http:", ignoreCase = true) || argument.startsWith("https:", ignoreCase = true)) {
                urlFound = true
                builder.url(argument)
            } else if (!argument.startsWith("-") && !urlFound) {
                urlFound = true
                builder.url(argument)
            }
        }
    }

    companion object {

        fun parse(commandString: String): CurlCommand {
            val arguments = CommandParser.parseCommand(commandString)
            val curlParser = CurlParser(arguments)
            return curlParser.builder.build()
        }

        fun isSupportedOption(option: String) =
            option in SUPPORTED_OPTIONS

        private val SUPPORTED_OPTIONS = setOf(
            "-X",
            "--request",
            "-x",
            "--proxy",
            "-H",
            "--header",
            "-d",
            "--data",
            "--data-binary",
            "--data-urlencode",
            "--data-raw",
            "--digest",
            "-F",
            "--form",
            "-m",
            "--max-time",
            "--connect-timeout",
            "-u",
            "--user",
            "-A",
            "--user-agent",
            "--url",
            "-e",
            "--referer",
            "-G",
            "--get",
            "-k",
            "--insecure",
            "-s",
            "--silent",
            "--head",
            "-I",
            "-4",
            "--ipv4",
            "-6",
            "--ipv6",
        )
    }
}
